frontend/apps/reader/modules/readerbookmark.lua

```diff
@@ -196,248 +219,359 @@ function ReaderBookmark:addToMainMenu(menu_items)
-                    b.pos1.page = b.page
-                end
-                self.ui.document.configurable.text_wrap = 0 -- native positions
-                -- sort start and end positions of each highlight
-                local compare_pos, a_start, a_end, b_start, b_end, result
-                compare_pos = self.ui.document:comparePositions(a.pos0, a.pos1) > 0
-                a_start = compare_pos and a.pos0 or a.pos1
-                a_end = compare_pos and a.pos1 or a.pos0
-                compare_pos = self.ui.document:comparePositions(b.pos0, b.pos1) > 0
-                b_start = compare_pos and b.pos0 or b.pos1
-                b_end = compare_pos and b.pos1 or b.pos0
-                -- compare start positions
-                compare_pos = self.ui.document:comparePositions(a_start, b_start)
-                if compare_pos == 0 then -- both highlights with the same start, compare ends
-                    result = self.ui.document:comparePositions(a_end, b_end) < 0
-                else
-                    result = compare_pos < 0
-                end
-                self.ui.document.configurable.text_wrap = is_reflow -- restore reflow mode
-                return result
-            end
-            return a.highlighted -- have page bookmarks before highlights
+function ReaderBookmark:toggleBookmark(pageno)
+    local pn_or_xp
+    if pageno then
+        if self.ui.rolling then
+            pn_or_xp = self.ui.document:getPageXPointer(pageno)
+        else
+            pn_or_xp = pageno
         end
-        return a.page > b.page
     else
-        local a_page = self.ui.document:getPageFromXPointer(a.page)
-        local b_page = self.ui.document:getPageFromXPointer(b.page)
-        if a_page == b_page then -- both bookmarks in the same page
-            local compare_xp = self.ui.document:compareXPointers(a.page, b.page)
-            if compare_xp then
-                if compare_xp == 0 then -- both bookmarks with the same start
-                    if a.highlighted and b.highlighted then -- both are highlights, compare ends
-                        compare_xp = self.ui.document:compareXPointers(a.pos1, b.pos1)
-                        if compare_xp then
-                            return compare_xp < 0
-                        end
-                        logger.warn("Invalid xpointer in highlight:", a.pos1, b.pos1)
-                        return
-                    end
-                    return a.highlighted -- have page bookmarks before highlights
-                end
-                return compare_xp < 0
-            end
-            -- if compare_xp is nil, some xpointer is invalid and will be sorted first to page 1
-            logger.warn("Invalid xpointer in highlight:", a.page, b.page)
+        pn_or_xp = self:getCurrentPageNumber()
+    end
+    local index = self:getDogearBookmarkIndex(pn_or_xp)
+    if index then
+        local item = table.remove(self.ui.annotation.annotations, index)
```

Comment:
Again a bit hard to review around here.
Just a feeling: this bit of hacking into self.ui.annotation.annotations here, shouldn't it be a service/method provided by ReaderAnnotation so we don't have to know about the annotation structure here?
Below, you use annotation:addItem(), so I would expect here to use annotation:removeItem().
(Just mentionning that, not insisting, this all looks quite painful :))